### DISCUSSION TITLE
Keycloak 25.0.2 not running over HTTP on Azure Container App

### DISCUSSION THREAD

**Discussion by jesuscorral** [keycloak/keycloak#32242] [Archive]
We are deploying Keycloak 25.0.2 in Azure Container App, and we want to run Keycloak over HTTP (without certificate) but trying to access to URL exposed by Azure Container App with HTTPS (https://keycloak-server--u9hs5rq.proudrock-baec4025.northeurope.azurecontainerapps.io/) we are getting the error showed in the screenshot. 

As we can see in the Network tab is failing calling to "[keycloak-server--u9hs5rq.proudrock-baec4025.northeurope.azurecontainerapps.io/resources/master/admin/en](https://keycloak-server--u9hs5rq.proudrock-baec4025.northeurope.azurecontainerapps.io/resources/master/admin/en)" but if we paste directly this url in the browser I getting the result without issues.

![image](https://github.com/user-attachments/assets/ceab7863-112d-4ec6-b57e-925ac5f73b0d)

**Realm settings configuration**:
	- General:
		○ Require SSL: None
	- Security defenses:
		○ X-Frame-Options: SAMEORIGIN
		○ Content-Security-Policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';
		○ X-XSS-Protection: 1; mode=block
                    HTTP Strict Transport Security (HSTS): max-age=31536000; includeSubDomains

**We are creating Docker image with this Dockerfile:**
```

FROM quay.io/keycloak/keycloak:25.0.2 as builder
# Configurate database
ENV KC_DB=mssql
ENV KC_DB_URL="jdbc:sqlserver://keycloaktests.database.windows.net:1433;database=keycloak"
ENV KC_DB_USERNAME=***
ENV KC_DB_PASSWORD=****
# Keycloak configuration
ENV KEYCLOAK_ADMIN=**
ENV KEYCLOAK_ADMIN_PASSWORD=**

ENV KC_HOSTNAME_STRICT=false
ENV KC_HOSTNAME_STRICT_HTTPS=false
ENV KC_HTTP_ENABLED=true
ENV KC_PROXY_HEADERS=forwarded 
ENV KC_HTTPS_CLIENT_AUTH=none
ENTRYPOINT ["/opt/keycloak/bin/kc.sh", "start"]
```

**Azure Container App Ingress Configuration:**

![image](https://github.com/user-attachments/assets/b8549739-873e-4ed4-8652-c5025e165a63)

**Comment by codespearhead** [comment]
Try the suggested solution from https://github.com/keycloak/keycloak/discussions/28279 .

**Reply by jesuscorral** [reply]
I've already tried that solution, but I'm getting the error yet. Indeed in my dockerfile you can see that I hace implemented the potential solution from the other post. Thanks

**Reply by codespearhead** [reply]
Try again with the following lines removed from your Dockerfile (ie. using only the env vars in the referred answer [2]):

ENV KC_HOSTNAME_STRICT=false
ENV KC_HOSTNAME_STRICT_HTTPS=false
ENV KC_HTTP_ENABLED=true
ENV KC_HTTPS_CLIENT_AUTH=none

By the way, you don't need to create a custom Docker image just to add env vars (see [1]).

The link in the referred answer [2] talks about a very similar error tackled in section "Error case 1" of that article, which I believe applies to you as well, hence the removed lines from the Dockerfile.

[1] https://learn.microsoft.com/en-us/azure/container-instances/container-instances-environment-variables#azure-portal-example
[2] https://alonewolf24.medium.com/how-to-deploy-keycloak-in-azure-app-services-in-docker-f13da34b92e6

**Reply by jesuscorral** [reply]
I've tried your solution, running directly Keycloak image in Azure container app and removing the following lines:

ENV KC_HOSTNAME_STRICT=false
ENV KC_HOSTNAME_STRICT_HTTPS=false
ENV KC_HTTP_ENABLED=true
ENV KC_HTTPS_CLIENT_AUTH=none

**First attemp:**
 - Deploy Keycloak 25.0.2 directly from Azure Container App with these environment variables:
      KEYCLOAK_ADMIN admin
      KEYCLOAK_ADMIN_PASSWORD admin
      KC_DB mssql
      KC_DB_URL jdbc:sqlserver://keycloaktests.database.windows.net:1433;database=keycloak
      KC_DB_USERNAME ***
      KC_DB_PASSWORD ***
      KC_PROXY_HEADERS forwarded
 
      And command => Command /opt/keycloak/bin/kc.sh, start 

      **Result**: The revision can't be activated and I get this error from Log "Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode."

**Second attemp:**
   - I've used same list of environment variables but changing the command to run keycloak in dev mod with
     Command /opt/keycloak/bin/kc.sh, start-dev

     **Result**: The revision is activated but I get the same error than initially:  
     ![image](https://github.com/user-attachments/assets/d99b1f0c-3d83-464e-abba-d8b3b2f19e68)

**Third attemp:**
  - I've use a new database in case that I had some bad configuration saved on database but it doesn't works either

**Reply by codespearhead** [reply]
Try the latest version of Keycloak (25.0.4). If that doesn't work, run it without any en vars at all and check if any errors is thrown.

**Reply by jesuscorral** [reply]
I've update the image to latest one 25.0.4.
- Removing all env variables I get this expected error because I need at least  KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD 
  ![image](https://github.com/user-attachments/assets/b951cd47-be07-490b-a9c1-f67f737b3354)

- Removing all env variables except KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD I get the same error. Trying to get language file ussing HTTP instead use HTTPS I get the same error. 
  
![image](https://github.com/user-attachments/assets/9df6dd48-c918-440c-a9e4-9e2b6bd664ee)

**Reply by jesuscorral** [reply]
@codespearhead thanks for you help, finally I found the issue. I've added this env variable **KC_HOSTNAME_BACKCHANNEL_DYNAMIC = true** and now is working like a charm deployed in Azure Container App without SSL certificate.

**Reply by codespearhead** [reply]
Thanks for sharing your solution!

Unfortunately, I no longer have an Azure account to investigate it further, and the current workaround might leave your vulnerable as per the documentation:

https://github.com/keycloak/keycloak/blob/776a491989ebc9be435dea0cb57943e2ad6b65cf/docs/guides/server/hostname.adoc?plain=1#L11-L17

Your initial settings (`KC_HOSTNAME_STRICT=false` and `KC_PROXY_HEADERS=forwarded`) should've worked:

https://github.com/keycloak/keycloak/blob/776a491989ebc9be435dea0cb57943e2ad6b65cf/docs/guides/server/hostname.adoc?plain=1#L49-L63

The warning [on that very page](https://www.keycloak.org/server/hostname) might be a clue to what has changed in Azure's internals since I last tried it:

> WARNING: If either `forwarded` or `xforwarded` is selected, make sure your reverse proxy properly sets and overwrites the `Forwarded` or `X-Forwarded-*` headers respectively. To set these headers, consult the documentation for your reverse proxy. Misconfiguration will leave {project_name} exposed to security vulnerabilities.

Either way, I think it's important to keep this discussion open to raise awareness about this issue.

**Reply by sachingaikwad123** [reply]
I am facing this same problem and I am not able to solve with it given solution. How to fix this? I think I am missing some environment variable or some other setting?

Here is what I have in env variables:

KC_HOSTNAME_STRICT = false
KC_PROXY_HEADERS = forwarded

![image](https://github.com/user-attachments/assets/b4ce4c9a-2025-4e05-a7ef-7984879999f5)

**Reply by sachingaikwad123** [reply]
I added this environment variable:

KC_HOSTNAME_BACKCHANNEL_DYNAMIC = true

Now, I am getting following error when starting keycloak container (from logs):

ERROR: Unexpected error when starting the server in (development) mode
ERROR: Failed to start quarkus
ERROR: hostname-backchannel-dynamic must be set to false when no hostname is provided

**Reply by sachingaikwad123** [reply]
I am not setting these env vars. Do I need to set these?

ENV KC_HOSTNAME_STRICT=false
ENV KC_HOSTNAME_STRICT_HTTPS=false
ENV KC_HTTPS_CLIENT_AUTH=none

**Reply by codespearhead** [reply]
Insecure connections should not be allowed in production. Keycloak will refuse to do so unless you're in development mode. That's probably what's causing the issue.

If that doesn't work, try passing `KC_PROXY_HEADERS=xforwarded` instead of `KC_PROXY_HEADERS=forwarded`.

**Reply by sachingaikwad123** [reply]
Sorry, I forgot to mention that I am using keycloak in development mode right now. I want to do some functional testing first.

Command: /opt/keycloak/bin/kc.sh, start-dev

I am trying now with KC_PROXY_HEADERS=xforwarded

**Reply by codespearhead** [reply]
> I forgot to mention that I am using keycloak in development mode

Then you should try to access it via HTTP not HTTPS. Development mode allows access via HTTP without SSL/TLS.

> I am not setting these env vars. Do I need to set these?
> 
> ENV KC_HOSTNAME_STRICT=false ENV KC_HOSTNAME_STRICT_HTTPS=false ENV KC_HTTPS_CLIENT_AUTH=none

You're just passing the default value [1] to `KC_HTTPS_CLIENT_AUTH`, and `KC_HOSTNAME_STRICT_HTTPS` should not be set to false in production [2].

Try setting:

```bash
KC_BOOTSTRAP_ADMIN_PASSWORD=admin
KC_BOOTSTRAP_ADMIN_USERNAME=admin
KC_PROXY_HEADERS=xforwarded
KC_HOSTNAME=your_app_name.azurewebsites.net
```

Make sure you don't set anything else other than the variables above and the ones related to the database, if you're using an external one.

If you're running it in dev mode, then use `kc.sh start-dev`, as shown here [3].

***

[1] https://github.com/keycloak/keycloak/blob/10aca5552314d6e5c9226f3c6d0107861c38c693/quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java#L48-L53
[2] https://github.com/keycloak/keycloak/blob/10aca5552314d6e5c9226f3c6d0107861c38c693/quarkus/config-api/src/main/java/org/keycloak/config/HostnameV2Options.java#L21-L25
[3] https://www.keycloak.org/getting-started/getting-started-docker
